### PR TITLE
Add support for default values

### LINKED ISSUE #20: Better Property.value consumption in sort-destructuring-keys
Hey!
First of all, thank you for sharing your plugin. Quite a few rules are really useful in daily use case.
There is a minor flaw in the current implementation of `sort-destructuring-keys` rule though, though.
Such patterns are unsupported at the moment:
```js
const props = { a: 1, b: 2 }
const { b, a = b } = props
```

```js
const props = { a: 1, B: 2 }
const { z, b = z, a = z + b } = props
```

```js
const { z, c = 10, a = z() + c } = { z: () => 2 };
```

```js
const { b, a = (c = b) => c } = { b : 2 }
```

Basically, we could take nodes with type `AssignmentPattern` into consideration.
We would just list all keys from Identifiers located in valid nodes such as `ConditionalExpression`, `BinaryExpression`, `CallExpression`, `ArrowFunctionExpression`, `FunctionExpression` etc.
Next step would be to filter these keys and pick only the ones that are found within `ObjectPattern`.
Once that is done, we could use them in sorting function.
I would be more than happy to give it a shot, since it's something I am missing and would use in my projects.
What do you think about that? Do we want to support such a case?
If so, I will proceed with actual implementation.
Cheers.

### PR DESCRIPTION
**Pull request by P0lip**
Fixes #20 

This just an initial version, a grasp of new functionality.

TODO:

- [x] Support computed properties
- [x] Simplify the code and improve performance
- [x] Add more test cases
- [x] Support bodies of functions, but only in case they are immediately invoked

### CODE REVIEW

**Comment by P0lip** [comment]
Should we support Node 4?
If so, I will get rid of spread operator.

**Comment by wyze** [comment]
It is almost EOL, I will remove it from the Travis build.

**Comment by wyze** [comment]
Want to rebase on top of latest master or resolve the conflicts? Whichever you prefer.

**Comment by P0lip** [comment]
I will resolve the conflict :)

**Comment by P0lip** [comment]
Conflicts resolved, squashed all commits.
Could you please take a look at the test cases I added? Just to make sure we aren't missing anything.
I did try to include as many cases as possible, but also kept in mind to limit them to some reasonable amount.
If you can think of any useful scenario, let me know, please.
Thank you!

**Comment by wyze** [comment]
This is super interesting, what is used for comparison in the following code?

```js
const props = { z: () => 2 }
const { c = 10, z, a = z() + c } = props
```

**Comment by P0lip** [comment]
I am not sure if I understood you correctly, so my answer may be not the best one.
Anyway, I will try to explain the enhancement, as it may be a good answer to your question. :)
First of all, I don't make any assumptions on what `props` could be, since it's usually unknown during the runtime.
Obviously, such a case would fail
```js
const props = {}
const { c = 10, z, a = z() + c } = props
```
but in real scenario, props would be passed from some other place you can't really refer to, so it's up to a developer to either always include `z` in `props` or provide a default value for this property, i.e.
```js
const props = {}
const { c = 10, z = () => {}, a = z() + c } = props
```
-----
I always list all keys of each Identifier from `AssignmentPattern` (when it exists)
Then I remove those keys that are not present in the `ObjectPattern` (as you know, in our case it's an object you destruct from).
Once that is performed, I just remove prevent such a key from being placed before any dependent key.
For intance such an example is invalid
```
const props = { z: () => 2 }
const { z, c = 10, a = z() + c } = props
```
since `c` doesn't rely on z in any way, so it could be placed after `z`.

Is that clear?
If not, don't hesitate to ask. I will be really happy to explain any doubts or answer any questions you may have. :)
Cheers

**Comment by wyze** [comment]
I think I will still need some further explaination, sorry. I'm still not following, trying to take your explanation and look at the test cases.

```js
const props = {}
const { c = 10, z, a = ((x = c) => z)() } = props
```

The above is a test case, and it is in the `valid` array, and it passes. But I don't understand why `z` comes before `a` here.

**Comment by P0lip** [comment]
Keep in mind that those examples are quite difficult to reason about, since the code is like the most readable humanity has seen :)
I did it in purpose though, just to make sure that once autofix arrives, we won't break existing code.

Answering your question,
Basically it's because a ReferenceError would be thrown otherwise. 
Paste the following piece of code in your console
```js
{
  'use strict';
  const { c = 10, a = ((x = c) => z)(), z } = { z: 10 }
}
```
This throws `Uncaught ReferenceError: z is not defined`, while 
```js
{
const { c = 10, z, a = ((x = c) => z)() } = { z : 10 }
console.log(a); // logs 10
}
```
is perfectly fine.

There is one edge case, though, that happens when function is not immediately invoked.
This is considered as a valid example, since z won't be accessed immediately.
```js
{
const { c = 10, a = ((x = c) => z), z } = { z : 15 }
const b = a();
console.log(b); // logs 15
}
```
You must call that `a` function and once you do that, `z` will be already there :)


In other words, we must have those keys defined before to prevent having Reference Errors.
Is that more clear now?
If not, I will be happy to discuss it further :)

**Comment by wyze** [comment]
Thank you again for your wonderful, detailed responses. I completely understand now. I was contemplating putting this as an option, what are your thoughts on that?

It seems like the messages could get confusing if you had something like:

```js
const { a = z, z = 10 } = {}
```

I would assume it would say it would want `z`, before `a`, correct?

**Comment by P0lip** [comment]
Yes, we could definitely put it behind an option, as it's something optional and wouldn't be so widely used I believe.
Regarding provided case - Unfortunately, it wouldn't, but ESLint has a possibility to detect it (`no-use-before-define` rule).
I added just a possibility to allow them be placed before.

Do you think we should support it? It's doable and may be an interesting case as well. I would also put it behind an option.

File: lib/rules/sort-destructuring-keys.js [hunk 1]
```diff
@@ -54,6 +56,65 @@ const getOrder = ( order, modifier ) => {
 const getKeyName = node =>
   node.type === 'ExperimentalRestProperty' ? node.argument.name : node.key.name
 
+/**
+ * List all identifiers' name for given node
+ *
+ * @param {ASTNode} node - The property node.
+ * @returns {Array<String>} identifiers - List of identifiers
+ */
+const getIdentifiers = node => {
```
Comment (wyze) [inline lib/rules/sort-destructuring-keys.js]: Move this to a separate file in `helpers` so we can remove the `eslint-disable` comment at the top of the file.

File: lib/rules/sort-destructuring-keys.js [hunk 2]
```diff
@@ -82,9 +143,15 @@ module.exports = {
...
-        const keys = properties
+        const filteredProperties = properties
           .filter(prop => !prop.computed && prop.type !== 'ExperimentalRestProperty')
-          .map(prop => getKeyName(prop))
+        const allKeys = filteredProperties.map(getKeyName)
+        const keys = filteredProperties
+          .map(( prop, i ) => [
+            allKeys[i],
+            getKeyValue(prop)
+              .filter(Array.prototype.includes, allKeys),
```
Comment (wyze) [inline lib/rules/sort-destructuring-keys.js]: Nice!

Comment (P0lip) [reply lib/rules/sort-destructuring-keys.js]: Dang, this won't work properly in some cases, since `Array.prototype.includes` takes index as a second argument... and in this particular case it will receive index, since `filter` does have an index as its second argument...
Will need to fix that as well.

**Comment by P0lip** [comment]
Changes applied :)

**Comment by P0lip** [comment]
Shall I apply any other changes or can we proceed?

**Comment by wyze** [comment]
Sorry, had a busy work week! Thanks for putting up with me. :)

**Comment by P0lip** [comment]
Np, thanks for the merge! :)
